src/spark_rapids_ml/common/cuml_context.py

```diff
@@ -13,57 +13,142 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+import asyncio
 import base64
-from typing import Any, Optional
+import json
+import os
+from typing import Any, List, Optional, Tuple
 
+import psutil
 from pylibraft.common import Handle
 from pyspark import BarrierTaskContext
-from raft_dask.common.comms_utils import inject_comms_on_handle_coll_only
+from raft_dask.common import UCX
+from raft_dask.common.comms_utils import (
+    inject_comms_on_handle,
+    inject_comms_on_handle_coll_only,
+)
 from raft_dask.common.nccl import nccl
+from ucp import Endpoint
 
 
 class CumlContext:
     def __init__(
-        self, rank: int, nranks: int, context: BarrierTaskContext, enable: bool
+        self,
+        rank: int,
+        nranks: int,
+        context: BarrierTaskContext,
+        enable: bool,
+        require_ucx: bool = False,
     ) -> None:
         """
         Initialize the nccl unique id for workers.
 
         1. get the nccl unique id for worker 0
         2. do all gather for all the workers to get the nccl unique uid.
+        3. if require_ucx is true, initialize ucx and inject ucx together with nccl into a handle
         """
         self.enable = enable
         self._handle: Optional[Handle] = None
-        if enable:
-            nccl_uid = ""
-            if context.partitionId() == 0:
-                nccl_uid = base64.b64encode(nccl.get_unique_id()).decode("utf-8")
-            nccl_uids = context.allGather(nccl_uid)
+        if not enable:
+            return
+
+        self._rank = rank
+        self._nranks = nranks
+        self._require_ucx = require_ucx
+
+        self._handle = Handle(n_streams=0)
+        self._nccl_comm: Optional[nccl] = None
+        self._nccl_unique_id = None
+        self._ucx: Optional[UCX] = None
+        self._ucx_port = None
+        self._ucx_eps = None
+
+        nccl_uid = ""
+        if context.partitionId() == 0:
+            nccl_uid = base64.b64encode(nccl.get_unique_id()).decode("utf-8")
 
+        if self._require_ucx is False:
+            nccl_uids = context.allGather(nccl_uid)
             self._nccl_unique_id = base64.b64decode(nccl_uids[0])
-            self._rank = rank
-            self._nranks = nranks
-            self._handle = Handle(n_streams=0)
-            self._nccl_comm: Optional[nccl] = None
+        else:
+            tasks = context.getTaskInfos()
+            ips = [task.address.split(":")[0] for task in tasks]
+
+            # set environmental variables according to https://github.com/rapidsai/ucx-py
+            # the code occasionally run fail without setting the variables
+            my_ip = ips[self._rank]
+            my_ifname = CumlContext.get_ifname_from_ip(my_ip)
+            os.environ["UCX_TLS"] = "tcp,cuda_copy,cuda_ipc"
+            os.environ["UCXPY_IFNAME"] = my_ifname
```

Comment:
We'll have to figure how to make this more flexible to take advantage of higher speed interconnects on multi-gpu nodes.

Reply:
created a ticket.